### ISSUE TITLE
Authenticating for every request is not viable in most production applications

### ISSUE THREAD

**Issue by timrobinson33** [MicrosoftDocs/powerapps-docs#1470]
I am concerned that the top paragraph describes the code samples here as "re-usable" implying that you could simply copy and paste them into a production application, and I'm sure many pushed-for-time developers would do so without looking too closely at it.

However, the code causes the authentication to be re-executed for every request. Testing this on my home broadband, the authentication alone typically takes 250ms whereas a typical web api call with an existing token takes only around 50ms. therefore this strategy of authenticating every time is reducing the response time of the application by a factor of 5 unnecessarily. If the application were running on a cloud server with less latency, I would imagine the discrepancy would be even larger.

In addition, use of `new HttpClient` is now discouraged, so it would be useful to update the sample to integrate with HttpClientFactory

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 6da4c6b5-4ca0-0138-ea99-47e40feaf908
* Version Independent ID: 6bd64a3c-bdf4-a8dd-a6d0-a607a444a30f
* Content: [Enhanced quick Start(Common Data Service) - Power Apps](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/enhanced-quick-start)
* Content Source: [powerapps-docs/developer/common-data-service/webapi/enhanced-quick-start.md](https://github.com/MicrosoftDocs/powerapps-docs/blob/live/powerapps-docs/developer/common-data-service/webapi/enhanced-quick-start.md)
* Service: **powerapps**
* GitHub Login: @JimDaly
* Microsoft Alias: **jdaly**

**Comment by JimDaly** [comment]
@Tim -
The code includes the token with each request, but it doesn't actually acquire a new token each time.
The ADAL library manages this and doesn't actually get a new token each time. It is only refreshed when it is needed.

This is the pattern recommended by ADAL. 

As far has using HttpClientFactory, if this were a .NET Core sample that would be an option. But this is a basic quick start using .NET Framework. We are trying to keep things simple here.

Take a look at samples using this CDSWebAPIService class
[Web API CDSWebApiService class Sample (C#)](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/samples/cdswebapiservice)

Still .NET Framework, but they show a better pattern for managing the use of the HttpClient.

Then look at this new tutorial [Tutorial: Create an ASP.NET Core Blazor WebAssembly App using Common Data Service](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/walkthrough-blazor-webassembly-single-tenant)

Here you finally see HttpClientFactory and authentication with MSAL via the Blazor app.

There are many ways to use HttpClient with different options. We are trying to show a simple authentication example with basic best practices.

I will agree that we ought to get newer samples highlighting .NET Core and leveraging best practices for HttpClientFactory. But we can't cover all aspects in this rapidly changing space.

**Comment by timrobinson33** [comment]
Hi Jim and thanks for taking the time to reply. I didn't copy it verbatim but I thought mine was esentially the same and it was definitely re-authenticating every time. I'll take a closer look

**Comment by timrobinson33** [comment]
Hi Jim,

I have now tested this with exactly the code from the sample, and I'm not seeing what I think you're saying I should be seeing. My test was to run the WhoAmI function in an infinite loop in the main() method.

If I put the SampleHelpers.GetHttpClient call inside the loop, then each iteration of the loop takes about 150ms and I can see a ton of debug output where it's re-authenticating each time.

If I put the SampleHelpers.GetHttpClient call outside of the loop, then each iteration of the loop takes 50ms because it's not re-authenticating, but...  if I then put a `sleep(2000000)` (i.e. over 30 minutes) inside the loop, then the subsequent WhoAmI call will fail because it's trying to use a stale token.

It seems to me that in order to use the API efficiently in a scenario where there may be over 30 minutes between API calls, there must be some logic somewhere that says:

```
if the previous token has expired
   generate a new one
otherwise 
  use the same one".
```
 I don't see anything in the sample, or any evidence of anything inside ADAL which is doing this.

The text of the article says "...call the AcquireToken method every time an operation is performed so you don't need to explicitly manage token expiration". This implies to me that the author was expecting you to instantiate a new HttpClient for each request, i.e. re-authenticate every time, and this is what I was originally complaining about.

**Comment by JimDaly** [comment]
@timrobinson33  -
This topic is intended to help people understand how this pattern is applied in other samples. I'm not totally clear on what steps you are making to get the results you report, but I would recommend that you look at the next steps:
1. The other [Web API Data operations Samples (C#)](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/web-api-samples-csharp) which use this same pattern to make multiple calls while re-using the same HttpClient.  Certainly not expecting people to instantiate a new HttpClient for every request. Do you see the same issue there?
1. The samples that use the [Web API CDSWebApiService class Sample (C#)](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/samples/cdswebapiservice) 

We are in the process of retiring the older samples in favor of samples that use the CDSWebApiService because it includes better practices for managing a reusable HttpClient and include code to manage service protection limits.

This Enhanced quick start example is still just a quick start, building up from an even simpler experience describe in [Quick Start: Web API sample (C#)](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/quick-start-console-app-csharp)

**Comment by JimDaly** [comment]
@phecke 
Would you mind reviewing what @timrobinson33 is reporting here?
I know you have spent considerable time in this space.

**Comment by timrobinson33** [comment]
Hi Jim,

I've played with this some more and I think I have got to the bottom of it.

The CDSWebApiService class does call AcquireTokenAsync every time you make a request, but only the first call is slow; every subsequent time it is instantaneous As you said, there are some smarts inside ADAL to cache the returned token.

The sample on the "enhanced quick start" page I was originally complaining about contains the same logic as CDSWebApiService but for some reason it takes up to 100ms to return the token the second and subsequent time, **even though it's not re-authenticating**. As far as I can tell, the difference is down to the ADAL ADAL 2.29 version used in that sample.

In my case I am actually using MSAL which is slightly different. Whereas ADAL is caching the token in a global variable, MSAL requires you to construct a ConfidentialClientApplicationBuilder object and it seems to be here where the token is cached, so if you construct a new ConfidentialClientApplicationBuilder each time, then it will re-authenticate each time, but if you reuse the  ConfidentialClientApplicationBuilder, you can call AcquireTokenForClient every time and it won't re-authenticate.

So in conclusion, I withdraw my allegation that your sample is re-authenticating each time, but the sample is still slow because ADAL 2.29 takes a long time to return the cached token.

**Comment by JimDaly** [comment]
Thanks for letting us know. Much appreciated.
I'm also working on getting our samples to work with MSAL (and .NET Core). If you have any examples to share, I would certainly be very happy to leverage what you have learned.

But I'll close this for now.

**Comment by timrobinson33** [comment]
Hi Jim,

I appreciate how difficult it must be for you guys to keep up with a landscape that's changing so fast. I don't think there's any meaningful example I could share because I found the transition from ADAL to MSAL to be pretty straightforward - The main takeaway is that you should create a single ConfidentialClientApplication on startup () and call AcquireTokenForClient every time you want to make a new request.
